src/mlpack/methods/ann/rbm_impl.hpp

```diff
@@ -0,0 +1,224 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_RBM_IMPL_HPP
+#define MLPACK_METHODS_ANN_RBM_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "rbm.hpp"
+
+#include <mlpack/core.hpp>
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/math/random.hpp>
+
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial neural networks. */ {
+
+
+template<typename InitializationRuleType, typename RBMPolicy>
+RBM<InitializationRuleType, RBMPolicy>::RBM(
+    arma::mat predictors,
+    InitializationRuleType initializeRule,
+    RBMPolicy rbmPolicy,
+    const size_t numSteps,
+    const size_t mSteps,
+    const bool useMonitoringCost,
+    const bool persistence):
+    rbmPolicy(rbmPolicy),
+    initializeRule(initializeRule),
+    numSteps(numSteps),
+    mSteps(mSteps),
+    useMonitoringCost(useMonitoringCost),
+    persistence(persistence),
+    reset(false)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::Reset()
+{
+  size_t weight = rbmPolicy.Parameters().n_elem;
+
+  positiveGradient.set_size(weight, 1);
+  negativeGradient.set_size(weight, 1);
+  negativeSamples.set_size(rbmPolicy.VisibleSize(), 1);
+  tempNegativeGradient.set_size(weight, 1);
+
+  parameter.set_size(weight, 1);
+  parameter.zeros();
+  positiveGradient.zeros();
+  negativeGradient.zeros();
+  tempNegativeGradient.zeros();
+  initializeRule.Initialize(parameter, parameter.n_elem, 1);
+
+  rbmPolicy.Parameters() = arma::mat(parameter.memptr(), weight, 1, false,
+      false);
+  rbmPolicy.Reset();
+  reset = true;
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+template<typename OptimizerType>
+void RBM<InitializationRuleType, RBMPolicy>::
+    Train(const arma::mat& predictors, OptimizerType& optimizer)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+
+  if (!reset)
+    Reset();
+
+  // Train the model.
+  Timer::Start("rbm_optimization");
+  optimizer.Optimize(*this, parameter);
+  Timer::Stop("rbm_optimization");
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+double RBM<InitializationRuleType, RBMPolicy>
+    ::FreeEnergy(arma::mat&& input)
+{
+  return rbmPolicy.FreeEnergy(std::move(input));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+double RBM<InitializationRuleType, RBMPolicy>::Evaluate(
+    const arma::mat& /* parameters*/, const size_t i)
+{
+  if (!useMonitoringCost)
+  {
+    // negative samples used here are from previous iteration
+    if (negativeSamples.is_empty())
+      Gibbs(std::move(predictors.col(i)), std::move(negativeSamples));
```

Comment:
The problem is tricky since `MiniBatchSGD` updates parameters before it calls `Evaluate()`. Hence, the last negative samples are outdated. I suggest to leave as is right now, this is not a high priority problem.